api/ResponseTypes.d.ts

```diff
@@ -0,0 +1,132 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+interface SearchResponse<TSource = unknown, TAggregations = unknown> {
+  took: number
+  timed_out: boolean
+  _scroll_id?: string
+  _shards: Shards
+  hits: {
+    total: {
+      value: number
+      relation: string
+    }
+    max_score: number
+    hits: Array<{
```

Comment:
@delvedor I'd suggest extract this inline object literal type into an interface. I in my code usually use not only the `Source` type, but the whole hit type because there is an id on that I'd like to use:
```ts
interface DocHit<TSource> {
    _id: string;
    _source: TSource;
}
```
If we don't extract this into an interface we won't be able to write the following code:

```ts
interface MyDocSource { foo: number }
// the type of hit should be DocHit<MyDocSource> but we don't have the DocHit type as per this PR right now
function processHit(hit: ??) {
    console.log("doc with id", hit._id, "has foo:", hit._source.foo);
}

const res = await client.search<MyDocSource>({ ... });

// notice that we cannot annotate the docHit param type with an interface name here to
res.body.hits.hits.forEach(docHit => processHit(docHit))
```

Reply:
+1 here. I hit this problem a couple of days ago. An example from Kibana codebase: 
https://github.com/elastic/kibana/blob/master/x-pack/plugins/logstash/server/types.ts#L11-L12